### ISSUE TITLE
2019-11-19 Comp engine error when all concentrations are NA for tc008 M4SDSS, tc018 M4SS

### ISSUE THREAD

**Issue by jhhughes256** [tensfeldt/openNCA#27]
### Error

When setting all concentrations to NA (changed in the data), an error occurs for tc010, 011, 012, 013 and 015 when otherwise unchanged (unchanged mct). No errors occur for tc014 and tc016 when otherwise unchanged (unchanged mct).

All test cases except tc013 provide the following error when setting all concentrations to NA **and** a partial AUC is asked for by the mct (besides partial AUCs starting from zero). Note that when unchanged, tc010, tc011 and tc015 all contain such partial AUCs.

```
 Error in value[[3L]](cond) : Error in tmp[[2]]: subscript out of bounds
For SDEID 2925168 
7. stop(paste0(e, "For SDEID ", unique(data_data[, map_data$SDEID])[i])) at run_M1_SD_computation.R#1701
6. value[[3L]](cond) 
5. tryCatchOne(expr, names, parentenv, handlers[[1L]]) 
4. tryCatchList(expr, classes, parentenv, handlers) 
3. tryCatch({
    tmp_df <- data_data[data_data[, map_data$SDEID] == unique(data_data[, 
        map_data$SDEID])[i], ]
    tmp_df <- tmp_df[order(tmp_df[, map_data$TIME]), ] ... at run_M1_SD_computation.R#820
2. run_M1_SD_computation(data = merged_data, map = map_data, method = method, 
    model_regex = model_regex, parameter_list = parameter_list, 
    return_list = return_list) at run_computation.R#2217
1. run_computation(data = d, map = mct, flag = flags, parameterset = parameterset) 
```

When left unchanged, tc012 gives the following error when all concentrations are set to NA:

```
 Error in value[[3L]](cond) : 
  Error in if (time[e] == t_last) {: missing value where TRUE/FALSE needed
For SDEID 24280050 
7. stop(paste0(e, "For SDEID ", unique(data_data[, map_data$SDEID])[i])) at run_M1_SS_computation.R#2608
6. value[[3L]](cond) 
5. tryCatchOne(expr, names, parentenv, handlers[[1L]]) 
4. tryCatchList(expr, classes, parentenv, handlers) 
3. tryCatch({
    if (comp_required[["DOSECi"]] || comp_required[["DOSEC"]]) {
        dose_ci <- list()
    } ... at run_M1_SS_computation.R#1020
2. run_M1_SS_computation(data = merged_data, map = map_data, method = method, 
    model_regex = model_regex, parameter_list = parameter_list, 
    return_list = return_list) at run_computation.R#2223
1. run_computation(data = d, map = mct, flag = flags, parameterset = parameterset) 
```

Unlike the other test cases, tc013 provides the following error when all concentrations to NA and does not appear to change when adding a partial AUC.

```
Error in value[[3L]](cond) : 
  Error in est_c_0$est_c0: $ operator is invalid for atomic vectors
For SDEID 10589286 
7. stop(paste0(e, "For SDEID ", unique(data_data[, map_data$SDEID])[i])) at run_M2_SD_computation.R#1663
6. value[[3L]](cond) 
5. tryCatchOne(expr, names, parentenv, handlers[[1L]]) 
4. tryCatchList(expr, classes, parentenv, handlers) 
3. tryCatch({
    tmp_df <- data_data[data_data[, map_data$SDEID] == unique(data_data[, 
        map_data$SDEID])[i], ]
    tmp_df <- tmp_df[order(tmp_df[, map_data$TIME]), ] ... at run_M2_SD_computation.R#805
2. run_M2_SD_computation(data = merged_data, map = map_data, method = method, 
    model_regex = model_regex, parameter_list = parameter_list, 
    return_list = return_list) at run_computation.R#2229
1. run_computation(data = d, map = mct, flag = flags, parameterset = parameterset) 
```

**Comment by jhhughes256** [comment]
### Update

As of a48a57bb3f this issue still occurs. It also occurs for the new testcase, tc071.

**Comment by rudraya** [comment]
This issue was addressed in 02b2470, please let me know if it still occurs.

**Comment by jhhughes256** [comment]
### Update

The "if (time[e] == t_last)" and "est_c_0$est_c0" errors still occur as of 02b2470. The "subscript out of bounds" error no longer occurs, which may be related to the closing of #6. 

The "if (time[e] == t_last)" error can be replicated using tc010, tc011, tc012 & tc015 with `d$PKCNCN <- rep(NA, nrow(d))`. This error does not occur for tc014, tc016 (with OPTIMIZEKEL == 0) or tc071. This error does occur for tc016 when OPTIMIZEKEL == 1.

The "est_c_0$est_c0" only occurs for tc013.

**Comment by tensfeldt** [comment]
#27 remains open with commit 96f6417

Note that the previous observation/test result that #27 didn't occur with tc014 was a testing error (PKCNCN was being set to NA but mct$CONC was set to a different variable). 

with commit 96f6417 this remains an issue with tc010, tc011, tc012, tc013, tc014, tc015, tc017, tc018

Note that this defect doesn't occur with tc016 nor tc071
Note with tc017 and tc018, there's a different error message reported than for the rest of the test cases

Note also aside from tc017 and tc018, tc016 and tc071 are the only test cases where FLGEXKEL is all set to 1, i.e. no terminal elimination phase calculations are requested.


## tc010

```
> results_list <- run_computation(data=d, map=mct, flag=flags, parameterset=parameterset)
Error in value[[3L]](cond) : 
  Error in if (time[e] == t_last) {: missing value where TRUE/FALSE needed
For SDEID 2925168
In addition: Warning messages:
1: In run_M1_SD_computation(data = merged_data, map = map_data, method = method,  :
  Flag 'FLGACCEPTTAU' cannot be computed if 'TAUi' is not provided
2: In run_M1_SD_computation(data = merged_data, map = map_data, method = method,  :
 
 Show Traceback
 
 Rerun with Debug
 Error in value[[3L]](cond) : 
  Error in if (time[e] == t_last) {: missing value where TRUE/FALSE needed
For SDEID 2925168 > r <- results_list$data_out
Error: object 'results_list' not found
```

# tc011

```
> results_list <- run_computation(data=d, map=mct, flag=flags, parameterset=parameterset)
Error in value[[3L]](cond) : 
  Error in if (time[e] == t_last) {: missing value where TRUE/FALSE needed
For SDEID 70745987
In addition: Warning message:
In run_M1_SS_computation(data = merged_data, map = map_data, method = method,  :
 
 Show Traceback
 
 Rerun with Debug
 Error in value[[3L]](cond) : 
  Error in if (time[e] == t_last) {: missing value where TRUE/FALSE needed
For SDEID 70745987 
```
## tc012

```
> results_list <- run_computation(data=d, map=mct, flag=flags, parameterset=parameterset)
Error in value[[3L]](cond) : 
  Error in if (time[e] == t_last) {: missing value where TRUE/FALSE needed
For SDEID 24280050
In addition: Warning message:
In run_M1_SS_computation(data = merged_data, map = map_data, method = method,  :
 
 Show Traceback
 
 Rerun with Debug
 Error in value[[3L]](cond) : 
  Error in if (time[e] == t_last) {: missing value where TRUE/FALSE needed
For SDEID 24280050 
```

## tc013

```
> results_list <- run_computation(data=d, map=mct, flag=flags, parameterset=parameterset)
Error in value[[3L]](cond) : 
  Error: $ operator is invalid for atomic vectors
For SDEID 10589286
In addition: Warning messages:
1: In run_M2_SD_computation(data = merged_data, map = map_data, method = method,  :
  Flag 'FLGACCEPTTAU' cannot be computed if 'TAUi' is not provided
2: In run_M2_SD_computation(data = merged_data, map = map_data, method = method,  :
 
 Show Traceback
 
 Rerun with Debug
 Error in value[[3L]](cond) : 
  Error: $ operator is invalid for atomic vectors
For SDEID 10589286 
```
## tc014 

```
> results_list <- run_computation(data=d, map=mct, flag=flags, parameterset=parameterset)
Error in value[[3L]](cond) : 
  Error in if (time[e] == t_last) {: missing value where TRUE/FALSE needed
For SDEID 50896291
In addition: Warning message:
In run_M2_SS_computation(data = merged_data, map = map_data, method = method,  :
 
 Show Traceback
 
 Rerun with Debug
 Error in value[[3L]](cond) : 
  Error in if (time[e] == t_last) {: missing value where TRUE/FALSE needed
For SDEID 50896291 
```
# tc015

```
> results_list <- run_computation(data=d, map=mct, flag=flags, parameterset=parameterset)
Error in value[[3L]](cond) : 
  Error in if (time[e] == t_last) {: missing value where TRUE/FALSE needed
For SDEID 12659123
In addition: Warning messages:
1: In run_M3_SD_computation(data = merged_data, map = map_data, method = method,  :
  Flag 'FLGACCEPTTAU' cannot be computed if 'TAUi' is not provided
2: In run_M3_SD_computation(data = merged_data, map = map_data, method = method,  :
 
 Show Traceback
 
 Rerun with Debug
 Error in value[[3L]](cond) : 
  Error in if (time[e] == t_last) {: missing value where TRUE/FALSE needed
For SDEID 12659123 
```

## tc017 has a different error message related to all concentrations missing

```
> results_list <- run_computation(data=d, map=mct, flag=flags, parameterset=parameterset)
Error in value[[3L]](cond) : 
  Error in ae(amt = amt, time = tmp_df[, map_data$TIME]): Error in ae: 'amt' is not a numeric vector
For SDEID 58098661
In addition: Warning messages:
1: In run_M4_SD_computation(data = merged_data, map = map_data, method = method,  :
  Flag 'FLGEMESIS' is not present in the dataset
2: In run_M4_SD_computation(data = merged_data, map = map_data, method = method,  :
  Flag 'FLGACCEPTPREDOSECRIT' is not present in the dataset
3: In run_M4_SD_computation(data = merged_data, map = map_data, method = method,  :
 
 Show Traceback
 
 Rerun with Debug
 Error in value[[3L]](cond) : 
  Error in ae(amt = amt, time = tmp_df[, map_data$TIME]): Error in ae: 'amt' is not a numeric vector
For SDEID 58098661 
```

## tc018

```
> results_list <- run_computation(data=d, map=mct, flag=flags, parameterset=parameterset)
Error in value[[3L]](cond) : 
  Error in ae(amt = amt, time = tmp_df[, map_data$TIME]): Error in ae: 'amt' is not a numeric vector
For SDEID 43456658
In addition: Warning messages:
1: In run_M4_SS_computation(data = merged_data, map = map_data, method = method,  :
  Flag 'FLGEMESIS' is not present in the dataset
2: In run_M4_SS_computation(data = merged_data, map = map_data, method = method,  :
  Flag 'FLGACCEPTPREDOSECRIT' is not present in the dataset
3: In run_M4_SS_computation(data = merged_data, map = map_data, method = method,  :
 
 Show Traceback
 
 Rerun with Debug
 Error in value[[3L]](cond) : 
  Error in ae(amt = amt, time = tmp_df[, map_data$TIME]): Error in ae: 'amt' is not a numeric vector
For SDEID 43456658 
```

## tc071

```
> results_list <- run_computation(data=d, map=mct, flag=flags, parameterset=parameterset)
unit_conversion : Unit Class 1 (Time) time_col:  TMAX1 TMIN1 TLAST TLAG KELTMLO KELTMHI THALF LASTTIME MRTEVIFO1 MRTEVIFP1 TAU1  parameters are scaled from  HR  to  HR  via scaling factor:  1 
unit_conversion : Unit Class 3 (Dose) dose_col:  DOSE1  parameters are scaled from  MG  to  MG  via scaling factor:  1 
unit_conversion : Unit Class 4 (Volume) volume_col:  VZFTAU1  parameters are scaled from  ML  to  L  via scaling factor:  0.001 
unit_conversion : Unit Class 5 (Amount/Volume) conc_col:  CMAX1 CMIN1 CLAST1 CTROUGH1 CTROUGHEND1 CAV1  parameters are scaled from  NG/ML  to  NG/ML  via scaling factor:  1 
unit_conversion : Unit Class 6 (1/Time) kel_col:  KEL  parameters are scaled from  1/HR  to  1/HR  via scaling factor:  1 
unit_conversion : Unit Class 7 (Volume/Time) cl_col:  CLFTAU1  parameters are scaled from  ML/HR  to  L/HR  via scaling factor:  0.001 
unit_conversion : Unit Class 8: (Amount.Time/Volume) auc_col:  AUCALL AUCLAST AUCLAST1 AUCLASTC1 AUCINFO AUCINFOC AUCINFP AUCINFPC AUCTAU1  parameters are scaled from  NG.HR/ML  to  NG.HR/ML  via scaling factor:  1 
unit_conversion : Unit Class 9: (Amount.Time.Time/Volume) aumc_col:  AUMCTAU1  parameters are scaled from  NG.HR.HR/ML  to  NG.HR.HR/ML  via scaling factor:  1 
unit_conversion : Unit Class 12: ([Amount/Volume]/Amount) concdn_col:  CMAXDN1  parameters are scaled from  NG/ML/MG  to  NG/ML/MG  via scaling factor:  1 
Warning messages:
1: In run_M1_SS_computation(data = merged_data, map = map_data, method = method,  :
  Flag 'FLGEMESIS' is not present in the dataset
2: In unit_conversion(data = data_data, map = map_data, result = computation_df,  :
  'AMOUNTU' #2# is not present in the dataset provided via 'map'
3: In unit_conversion(data = data_data, map = map_data, result = computation_df,  :
  'CONCU' and 'DOSEU' values provided via 'map' is not present in the dataset provided via 'data'
4: In unit_conversion(data = data_data, map = map_data, result = computation_df,  :
  'PKPTMU' and 'CONCU' #10# are not present in the dataset provided via 'map'
5: In unit_conversion(data = data_data, map = map_data, result = computation_df,  :
  'CONCU' and 'NORMBSU' value provided via 'map' #13# are not present in the dataset provided via 'data'
6: In unit_conversion(data = data_data, map = map_data, result = computation_df,  :
  'CONCU', 'PKPTMU and 'NORMBSU' #19# are not present in the dataset provided via 'map'
> table(flags$FLGEXKEL, exclude=c())

  1 
600 
```

**Comment by rudraya** [comment]
This issue was addressed in 02b2470, please let me know if it still occurs.

**Comment by tensfeldt** [comment]
Issue #27 is still occurring with commit 6aba420 with tc013. Not occurring with tc010, tc011, tc012, tc014, tc015, tc016, ***tc017***, ***tc018***

Note that while this specific issue doesn't occur with tc017 and tc018 there's an essential difference in how the code behaves. With the other testcases, with all concentrations set to NA, the estimated concentration dataset has the same # of rows as the # of records in the input flags dataset with FLGEXKEL==0. With tc017 however, the estimated concentration dataset is generated but has no records (it does have headers). The behavior must be consistent.

**Comment by rudraya** [comment]
This issue was addressed in 18821dc, please let me know if it still occurs.

**Comment by jhhughes256** [comment]
### Update

This issue no longer occurs for tc013 as of 18821dc. I don't currently have tc008_M4SDSS set up for testing in my environment, so I cannot confirm if it can still be replicated for that test case. 

I'll leave this open for you to close @tensfeldt , but I will remove the priority label. If you still experience the issue, please feel free to add back the priority label.

**Comment by tensfeldt** [comment]
Retested with comment 18821dc with the following output for tc008_M4SDSS

```
> results_list <- run_computation(data=d, map=mct, flag=flags, parameterset=parameterset)
Error in value[[3L]](cond) : 
  Error in if (is.null(conc1) || is.na(conc1)) {: missing value where TRUE/FALSE needed
For SDEID 225930823
In addition: Warning messages:
1: In run_M4_SD_computation(data = merged_data, map = map_data, method = method,  :
  Flag 'FLGEMESIS' is not present in the dataset
2: In run_M4_SD_computation(data = merged_data, map = map_data, method = method,  :
  Flag 'FLGACCEPTPREDOSECRIT' is not present in the dataset
3: In run_M4_SD_computation(data = merged_data, map = map_data, method = method,  :
 
 Show Traceback
 
 Rerun with Debug
 Error in value[[3L]](cond) : 
  Error in if (is.null(conc1) || is.na(conc1)) {: missing value where TRUE/FALSE needed
For SDEID 225930823 
```

**Comment by jhhughes256** [comment]
### Update

As of f98f92336f this issue now occurs for tc018 M4SS with the following error.

```
 Error in value[[3L]](cond) : 
  Error in `[<-.data.frame`(`*tmp*`, i, , value = c("43456658", NA, NA, : replacement has 45 items, need 47
For SDEID 43456658 
7. stop(paste0(e, "For SDEID ", unique(data_data[, map_data$SDEID])[i])) at run_M4_SS_computation.R#1507
6. value[[3L]](cond) 
5. tryCatchOne(expr, names, parentenv, handlers[[1L]]) 
4. tryCatchList(expr, classes, parentenv, handlers) 
3. tryCatch({
    if (comp_required[["DIi"]]) {
        di <- list()
    } ... at run_M4_SS_computation.R#688
2. run_M4_SS_computation(data = merged_data, map = map_data, method = method, 
    model_regex = model_regex, parameter_list = parameter_list, 
    return_list = return_list) at run_computation.R#2337
1. run_computation(data = d, map = mct, flag = flags, parameterset = parameterset) 
```

This issue currently applies to the following testcases:

- tc008 M4SDSS (prior to #67)
- tc018 M4SS (as of f98f92336f)

**Comment by rudraya** [comment]
If this issue was with respect to tc008 then I was addressed some fixes as of f98f923, please let me know if it still occurs.

**Comment by rudraya** [comment]
This issue was addressed in fce7a3c by addressing issue #67, please let me know if it still occurs.

**Comment by jhhughes256** [comment]
### Update

This issue can no longer be replicated for tc008 M4SD. Due to issue #86 tc018 M4SS can't be tested. Closing this issue for now, as it is not clear if #86 is due to NA values or something else.
